### PR TITLE
Feat/select billing frequency tapc 4663

### PR DESCRIPTION
**Pull request by ZyaCon**
## Description

Ticket Reference: #TAPC-4663

### CODE REVIEW

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts [hunk 1]
```diff
@@ -105,6 +105,10 @@ export type TPrice = {
   monthlyEquivalent: TPriceDetails | null;
 };
 
+/**
+ *  id: flavor_region_osType
+ *  Ex: 'B3-8_BHS1.PREPROD_linux'
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts]: `_price`is missing ;) 
ex: `c3-128-flex_RBX-A_windows_price`

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx [hunk 2]
```diff
@@ -48,6 +57,13 @@ export const CreateInstanceForm = () => {
     osType === 'windows' || distributionImageType === 'windows';
 
   const hasMultiMicroRegions = microRegions ? microRegions.length > 1 : false;
+
+  const pricingTypes = selectBillingTypes(deps)(
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx]: could you add a useMemo here to avoid extra computations?
also maybe harmonize `pricingTypes` with `selectBillingTypes`: `pricingTypes` => `billingTypes`

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import { TInstanceCreationForm } from '../CreateInstance.page';
+import {
+  Controller,
+  ControllerFieldState,
+  ControllerRenderProps,
+  FieldValues,
+  useFormContext,
+  UseFormStateReturn,
+  useWatch,
+} from 'react-hook-form';
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx:16]: you should remove unused imports here

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    name: [
+      'billingType',
+      'flavorId',
+      'distributionImageVersion.distributionImageVersionName',
+    ],
+  });
+
+  const handleSelectBilling = (billingType: BILLING_TYPE | null) => {
+    if (!billingType) return;
+    const pricingTypes = selectBillingTypes(deps)(
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: here you don't need `selectBillingTypes` you should deal with `billingTypes`

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+          {t('creation:pci_instance_creation_select_billing_choice_title')}
+        </Text>
+      </div>
+      <Controller
+        name="billingType"
+        control={control}
+        render={() => (
+          <RadioGroup
+            value={selectedBillingType}
+            onValueChange={({ value }) => handleSelectBilling(value)}
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: here you have a type mismatch : `handleSelectBilling` expects a `BILLING_TYPE` and value is a string ;)

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                onClick={() => updateBillingSelection(BILLING_TYPE.Hourly)}
+                selected={selectedBillingType === BILLING_TYPE.Hourly}
+              >
+                <Radio value="hourly">
+                  <div className="flex flex-col items-start gap-2">
+                    <RadioLabel className="flex items-center gap-x-4 text-lg font-bold text-[--ods-color-heading]">
+                      <RadioControl />
+                      <Text className="font-medium text-[--ods-color-heading]">
+                        {t('pci_instances_creation_select_billing_hourly')}
+                      </Text>
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: could you prefix all translations with its namespace? it will be safer if we add a new namespace in the future :)

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      <Controller
+        name="billingType"
+        control={control}
+        render={() => (
+          <RadioGroup
+            value={selectedBillingType}
+            onValueChange={({ value }) => handleSelectBilling(value)}
+          >
+            <div className="flex items-center space-x-4">
+              <PciCard
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: here to avoid code duplications you can iterate on `billingTypes`

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+export const selectBillingTypes: Reader<Deps, TSelectBillingData> = (deps) => {
+  return (projectId, flavorId, distributionImageVersionName) => {
+    if (!distributionImageVersionName || !flavorId) return [];
+
+    const { instancesCatalogPort } = deps;
+
+    const data = instancesCatalogPort.selectInstancesCatalog(projectId);
+    if (!data) return [];
+
+    const image = data.entities.images.byId.get(distributionImageVersionName); // a changer apres le merge
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: extra comment here ^^

Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: here you can directly deal with `distributionImageOsType` :)

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    const flavorPriceId = `${flavorId}_${image?.osType}_price`;
+
+    const periodPrices = data.entities.flavorPrices.byId.get(flavorPriceId)
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: may switch `periodPrices` with `prices` & vice versa?

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts (hunk 1 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts]: and also:  `flavor_region_osType` => d: `flavor_region_osType_price`

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx (hunk 2 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx]: `useMemo` ?

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 6 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: some are still missing

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 7 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: I cant' see the changes to map over it?

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    const data = instancesCatalogPort.selectInstancesCatalog(projectId);
+    if (!data) return [];
+
+    const image = data.entities.images.byId.get(osType);
+    if (!image) return [];
+
+    const flavorPriceId = `${flavorId}_${image.osType}_price`;
+
+    const price = data.entities.flavorPrices.byId.get(flavorPriceId)?.prices;
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: it should be plural here ;)

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    const periodPrice = price?.filter((t) => t.type !== 'licence') || [];
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: same here

**Review by lolottetheclash — CHANGES_REQUESTED** [review]

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts [hunk 12]
```diff
@@ -104,6 +104,10 @@ export type TPrice = {
   monthlyEquivalent: TPriceDetails | null;
 };
 
+/**
+ *  id: flavor_region_osType
+ *  Ex: 'B3-8_BHS1.PREPROD_linux_price'
+ */
```
Comment (aTurmo) [inline packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: Remark (non-blocking): I think you forgot to delete this

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  const { t } = useTranslation('creation');
+  const { control, setValue } = useFormContext<TInstanceCreationForm>();
+
+  const [selectedBillingType] = useWatch({
+    control,
+    name: [
+      'billingType',
+      'flavorId',
+      'distributionImageVersion.distributionImageVersionName',
+    ],
```
Comment (aTurmo) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: Question: why having 'flavorId' and 'distributionImageVersion.distributionImageVersionName' if you only get 'selectedBillingType' ?

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+          {t('creation:pci_instance_creation_select_billing_choice_title')}
+        </Text>
+      </div>
+      <Controller
+        name="billingType"
+        control={control}
+        render={() => (
+          <RadioGroup
+            value={selectedBillingType}
+            onValueChange={({ value }) =>
```
Comment (aTurmo) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: Question: I don't really get why there is an update on value here AND on each individual radio. Feels like it will trigger unecessary 'setValue'. And also is it necessary to hae the value in the `<RadioGroup>` if there is already a checkwith `selected={selectedBillingType === BILLING_TYPE.Hourly}` in each `<Radio>` ?

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/__tests__/billingTypesViewModel.spec.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    selectInstancesCatalog: vi
+      .fn()
+      .mockReturnValue(mockedInstancesCatalogEntity),
+  },
+  configurationPort: {
+    selectSshKeys: vi.fn(),
+  },
+};
+
+describe('selectBillingTypes ViewModel', () => {
```
Comment (aTurmo) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/__tests__/billingTypesViewModel.spec.ts:24]: Remark: It's missing tests when it should return an empty array.
Furthermore I don't see anywhere mention on the billing types.
Something that you should add is a clear indication of what the needed data are. So far they are hidden in mockedInstancesCatalogEntity and mockedInstancesCatalogDTO that you don't control, and if they chnge your test could break.
What you should have is data that looked like :
```javascript
mockEntity = {
   ...mockedInstancesCatalogEntity,
   flavorPrices: truc
   autre: matchin
}
```
And you override the data you need. This way the test is clearer and you controls the important data.

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts (hunk 12 above)
Comment (aTurmo) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: Or is this something you do in pci-instances ?

Comment (yaallco) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: This is an exemple of the TFlavorPrices value
I'm not sure to fully understand the question

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 13 above)
Comment (yaallco) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: I forgot to delete them, thanks

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 7 above)
Comment (yaallco) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: Thanks, I forgot to change that

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts (hunk 12 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: @aTurmo  This is to help understand the construction of the id :)

Comment (yaallco) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: Why should I delete this ?

Comment (aTurmo) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: I'm allergic to comment ^^ I think they generally pollute more than they help. But it's a case by case.
For example I think it's best to have good example in the test files.

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx [hunk 16]
```diff
@@ -48,6 +52,14 @@ export const CreateInstanceForm = () => {
     osType === 'windows' || distributionImageType === 'windows';
 
   const hasMultiMicroRegions = microRegions ? microRegions.length > 1 : false;
+
+  // const billingTypes = selectBillingTypes(deps)(projectId, flavorId, osType);
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/createInstanceForm/CreateInstanceForm.component.tsx]: comment here

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+              handleSelectBilling(value ? (value as BILLING_TYPE) : null)
+            }
+          >
+            <div className="flex items-center space-x-4">
+              {(Object.values(BILLING_TYPE) as BILLING_TYPE[]).map(
```
Comment (lolottetheclash) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: You should map on the `billingTypes` that you received from the viewModel used by the parent.
here you are using values from the enum, whereas you should use the actual values returned by the API

File: packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts (hunk 12 above)
Comment (lolottetheclash) [reply packages/manager/apps/pci-instances/src/domain/entities/instancesCatalog.ts:110]: I share your point of view, but here I think it's ok to let them as global code is not commented, it's just to expose at a glance the id format without having to check test files.

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+                        <RadioControl />
+                        <Text className="font-medium text-[--ods-color-heading]">
+                          {t(
+                            `creation:pci_instance_creation_select_billing_choice_${billing}`,
+                          )}
+                        </Text>
+                      </RadioLabel>
+                      <Text preset="caption">
+                        {t(
+                          'creation:pci_instance_creation_select_billing_choice_subtext',
```
Comment (aTurmo) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: Question: Is the 'choice_subtext' the same for every billing type ?

**Comment by fredericvilcot** [comment]
@ZyaCon Please consider squashing commits or rewording them as they are malformed (messages and body)
Thanks

File: packages/manager/apps/pci-instances/src/adapters/tanstack/instancesCatalog/right/dto.type.ts [hunk 19]
```diff
@@ -72,7 +72,7 @@ export type TFlavorRegionDTO = {
   tags: string[] | null;
 };
 
-type TPriceDetailsDTO = {
+export type TPriceDetailsDTO = {
```
Comment (fredericvilcot) [inline packages/manager/apps/pci-instances/src/adapters/tanstack/instancesCatalog/right/dto.type.ts]: Should this type being exported?

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+  Divider,
+  Radio,
+  RadioControl,
+  RadioGroup,
+  RadioLabel,
+  Text,
+} from '@ovhcloud/ods-react';
+import { useTranslation } from 'react-i18next';
+import { PciCard } from '@/components/pciCard/PciCard.component';
+import { TInstanceCreationForm } from '../CreateInstance.page';
```
Comment (fredericvilcot) [inline packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: A `component` should not rely on a `Page`, as they concern and responsibilities are not the same ;)

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+import { Reader } from '@/types/utils.type';
+
+type TSelectBillingData = (
+  projectId: string,
+  flavorId: string | null,
+  osType: string | null,
+) => BILLING_TYPE[];
+
+export const selectBillingTypes: Reader<Deps, TSelectBillingData> = (deps) => {
+  return (projectId, flavorId, osType) => {
```
Comment (fredericvilcot) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: Unnecessary `return` keyword :)


```suggestion
export const selectBillingTypes: Reader<Deps, TSelectBillingData> = (deps) => (projectId, flavorId, osType) => {
```

File: packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    const { instancesCatalogPort } = deps;
+
+    const data = instancesCatalogPort.selectInstancesCatalog(projectId);
+    if (!data) return [];
+
+    const flavorPriceId = `${flavorId}_${osType}_price`;
+
+    const prices = data.entities.flavorPrices.byId.get(flavorPriceId)?.prices;
+
+    const periodPrices = prices?.filter((t) => t.type !== 'licence') || [];
```
Comment (fredericvilcot) [inline packages/manager/apps/pci-instances/src/pages/instances/create/view-models/BillingTypesViewModel.ts]: What is `t` argument?

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 18 above)
Comment (yaallco) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: Thanks for the comment, I saw yesterday with the PO, new text has been added

File: packages/manager/apps/pci-instances/src/adapters/tanstack/instancesCatalog/right/dto.type.ts (hunk 19 above)
Comment (yaallco) [reply packages/manager/apps/pci-instances/src/adapters/tanstack/instancesCatalog/right/dto.type.ts]: Not anymore, thanks

File: packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx (hunk 20 above)
Comment (yaallco) [reply packages/manager/apps/pci-instances/src/pages/instances/create/components/BillingChoice.component.tsx]: That's true, there is a lot of changes to do on other components as well
